### PR TITLE
Change AssemblyName formatting not to require static table of KVPs

### PR DESCRIPTION
**Pull request by marek-safar**
to do loop with span operations to replace single characters

Expanded test coverage to actually cover this odd behaviour

### CODE REVIEW

File: src/libraries/System.Reflection/tests/AssemblyNameTests.cs [hunk 1]
```diff
@@ -34,6 +34,22 @@ public static IEnumerable<object[]> Names_TestData()
             yield return new object[] { "\uD83D\uDC3B\uD83D\uDC3B\uD83D\uDC3B\uD83D\uDC3B\uD83D\uDC3B", "\uD83D\uDC3B\uD83D\uDC3B\uD83D\uDC3B\uD83D\uDC3B\uD83D\uDC3B" };
         }
 
+        public static IEnumerable<object[]> Names_TestDataQuoted()
```
Comment (danmosemsft) [inline src/libraries/System.Reflection/tests/AssemblyNameTests.cs:37]: Why quoted -- would something like `Names_TestDataRequiresEscaping()` be better?

File: src/libraries/System.Reflection/tests/AssemblyNameTests.cs [hunk 2]
```diff
@@ -34,6 +34,22 @@ public static IEnumerable<object[]> Names_TestData()
...
+        {
+            yield return new object[] { " name ", "\" name \"" };
```
Comment (danmosemsft) [inline src/libraries/System.Reflection/tests/AssemblyNameTests.cs:39]: Did you miss the case of escaping a backslash?

**Review by danmoseley — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameFormatter.cs [hunk 3]
```diff
@@ -99,24 +99,32 @@ private static void AppendQuoted(this StringBuilder sb, string s)
...
-                    if (!(s[i] == escapeReplacement[0]))
+                    case '\\':
+                    case ',':
+                    case '=':
+                    case '\'':
+                    case '"':
+                        sb.Append('\\');
+                        break;
+                    case '\t':
+                        sb.Append('\\').Append('t');
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameFormatter.cs:112]: Is this faster than `sb.Append("\\t")`?

File: src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameFormatter.cs [hunk 4]
```diff
@@ -99,24 +99,32 @@ private static void AppendQuoted(this StringBuilder sb, string s)
...
                         continue;
-                    if ((s.Length - i) < escapeReplacement.Length)
+                    case '\n':
+                        sb.Append('\\').Append('n');
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameFormatter.cs:115]: Same question.  Wondering why we use the longer variant.

File: src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameFormatter.cs [hunk 5]
```diff
@@ -99,24 +99,32 @@ private static void AppendQuoted(this StringBuilder sb, string s)
...
-                        addedEscape = true;
-                    }
+#if TARGET_WINDOWS
+                    case '\r':
+                        if (++i < s.Length - 1 && s[i] == '\n')
+                            goto case '\n';
+
+                        --i;
+                        break;
+#endif
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameFormatter.cs:124]: Is this a change in behavior?  Seems like a \n on Windows (no proceeding \r) would have still resulted in a \r\n, but now it's only going to be a \n?

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameFormatter.cs:124]: And previously \r wasn't special-cased if it was on its own, was it?

File: src/libraries/System.Reflection/tests/AssemblyNameTests.cs (hunk 2 above)
Comment (marek-safar) [reply src/libraries/System.Reflection/tests/AssemblyNameTests.cs:39]: Included

File: src/libraries/System.Reflection/tests/AssemblyNameTests.cs (hunk 1 above)
Comment (marek-safar) [reply src/libraries/System.Reflection/tests/AssemblyNameTests.cs]: Updated the name

File: src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameFormatter.cs (hunk 3 above)
Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameFormatter.cs:112]: It's probably same now when Append is has a fast path for 2 characters string (updating)

File: src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameFormatter.cs (hunk 5 above)
Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameFormatter.cs:125]: Good point simplified it back

File: src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameFormatter.cs [hunk 6]
```diff
@@ -99,24 +99,36 @@ private static void AppendQuoted(this StringBuilder sb, string s)
...
+                    case '=':
+                    case '\'':
+                    case '"':
+                        sb.Append('\\');
+                        break;
+                    case '\t':
+                        sb.Append("\\t");
                         continue;
-                    if ((s.Length - i) < escapeReplacement.Length)
+#if TARGET_WINDOWS
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameFormatter.cs:114]: This Windows specific behavior sounds like a bug introduced in the managed rewrite of AssemblyName formatters. It does not really make sense for the AssemblyName escaping rules to differ between Windows and Unix.

I would recommend to change the managed formatter to do what the unmanaged formatter has been always doing: https://github.com/dotnet/runtime/blob/master/src/coreclr/src/binder/textualidentityparser.cpp#L740

**Review by jkotas — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameFormatter.cs (hunk 6 above)
Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameFormatter.cs]: Done.

**Review by stephentoub — APPROVED** [review]

**Pull request merged**